File: frontend/src/api/assistantContext.ts
import { api } from './client'

export type AssistantSurface =
  | 'setting_agent'
  | 'bootstrap_setting_agent'
  | 'plot_outline_agent'
  | 'workflow_intervention'
  | 'world_map_agent'
  | 'volume_planning'
  | 'outline_generation'

export interface AssistantContextSummary {
  packet_id?: string | null
  snapshot_id?: string | null
  snapshot_version?: number | null
  assistant_surface?: string | null
  invalidation_state: string
  force_reread: boolean
  history_reset_applied: boolean
  token_estimate: number
  selected_sections: Array<Record<string, any>>
  omitted_sections: Array<Record<string, any>>
  delta_count: number
  rebuilt_at?: string | null
}

export interface AssistantSessionResponse {
  session_id: string
  project_id: string
  assistant_surface: AssistantSurface | string
  mode: string
  status: string
  snapshot_id?: string | null
  snapshot_version?: number | null
  history_summary: string
  history_window: Array<Record<string, any>>
  pending_items: Array<Record<string, any>>
  context_cursor: Record<string, any>
  created_at?: string | null
  updated_at?: string | null
}

export interface AssistantMessageResponse {
  id: string
  session_id: string
  role: string
  content: string
  request_id?: string | null
  metadata: Record<string, any>
  packet_id?: string | null
  snapshot_id?: string | null
  created_at?: string | null
}

export interface AssistantHistoryResponse {
  session: AssistantSessionResponse
  messages: AssistantMessageResponse[]
  context_packet?: AssistantContextSummary | null
}

export interface AssistantSnapshotStatusResponse {
  project_id: string
  snapshot_id?: string | null
  snapshot_version?: number | null
  status: string
  source_revision_hash?: string | null
  content_hash?: string | null
  token_estimate: number
  stale_delta_count: number
  build_reason?: string | null
  built_at?: string | null
  error_message?: string | null
}

export interface AssistantHistoryResetResponse {
  success: boolean
  old_session_id: string
  new_session_id: string
  history_reset_at: string
  messages_deleted: number
  messages_archived: number
  pending_items_cleared: boolean
  snapshot_id?: string | null
  snapshot_version?: number | null
}

export interface AssistantForceRereadResponse {
  success: boolean
  project_id: string
  session_id?: string | null
  snapshot_id: string
  snapshot_version: number
  previous_snapshot_id?: string | null
  previous_snapshot_version?: number | null
  history_cleared: boolean
  force_reread: boolean
  packet_metadata: AssistantContextSummary
}

export async function createAssistantSession(
  projectId: string,
  data: {
    assistant_surface: AssistantSurface
    mode?: string
    session_id?: string
    scope?: Record<string, any>
    request_id?: string
  },
): Promise<AssistantSessionResponse> {
  return api.post(`/assistant-context/${projectId}/sessions`, data)
}

export async function getAssistantHistory(
  projectId: string,
  sessionId: string,
  limit = 200,
): Promise<AssistantHistoryResponse> {
  return api.get(`/assistant-context/${projectId}/sessions/${sessionId}/history`, { params: { limit } })
}

export async function resetAssistantHistory(
  projectId: string,
  sessionId: string,
  data: {
    assistant_surface: AssistantSurface
    mode?: string
    clear_pending_items?: boolean
    delete_message_records?: boolean
    reason?: string
    request_id?: string
  },
): Promise<AssistantHistoryResetResponse> {
  return api.post(`/assistant-context/${projectId}/sessions/${sessionId}/reset-history`, data)
}

export async function forceRereadProject(
  projectId: string,
  data: {
    assistant_surface: AssistantSurface
    session_id?: string
    mode?: string
    scope?: Record<string, any>
    clear_history?: boolean
    clear_pending_items?: boolean
    reason?: string
    request_id?: string
  },
): Promise<AssistantForceRereadResponse> {
  return api.post(`/assistant-context/${projectId}/force-reread`, data)
}

export async function resetAndReread(
  projectId: string,
  sessionId: string,
  data: {
    assistant_surface: AssistantSurface
    mode?: string
    scope?: Record<string, any>
    clear_pending_items?: boolean
    reason?: string
    request_id?: string
  },
): Promise<AssistantForceRereadResponse> {
  return api.post(`/assistant-context/${projectId}/sessions/${sessionId}/reset-and-reread`, {
    ...data,
    session_id: sessionId,
    clear_history: true,
  })
}

export async function getLatestSnapshot(projectId: string): Promise<AssistantSnapshotStatusResponse> {
  return api.get(`/assistant-context/${projectId}/snapshots/latest`)
}

export async function getContextPacket(projectId: string, packetId: string): Promise<Record<string, any>> {
  return api.get(`/assistant-context/${projectId}/packets/${packetId}`)
}

File: app/models/assistant_context.py
"""Unified assistant context fabric models."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import uuid4

from pydantic import BaseModel, Field


class AssistantSurface(str, Enum):
    SETTING_AGENT = "setting_agent"
    BOOTSTRAP_SETTING_AGENT = "bootstrap_setting_agent"
    PLOT_OUTLINE_AGENT = "plot_outline_agent"
    WORKFLOW_INTERVENTION = "workflow_intervention"
    WORLD_MAP_AGENT = "world_map_agent"
    VOLUME_PLANNING = "volume_planning"
    OUTLINE_GENERATION = "outline_generation"


class AssistantSessionStatus(str, Enum):
    ACTIVE = "active"
    ARCHIVED = "archived"
    RESET = "reset"


class SnapshotStatus(str, Enum):
    BUILDING = "building"
    READY = "ready"
    STALE = "stale"
    FAILED = "failed"


class SnapshotBuildReason(str, Enum):
    INITIAL = "initial"
    DELTA_THRESHOLD = "delta_threshold"
    MANUAL_FORCE_REBUILD = "manual_force_rebuild"
    SCHEMA_MIGRATION = "schema_migration"
    PROJECT_IMPORT = "project_import"


class DeltaOperation(str, Enum):
    CREATE = "create"
    UPDATE = "update"
    DELETE = "delete"
    RESTORE = "restore"


class PacketInvalidationState(str, Enum):
    FRESH = "fresh"
    DELTA_APPLIED = "delta_applied"
    STALE_REBUILD_REQUIRED = "stale_rebuild_required"
    FORCE_REBUILT = "force_rebuilt"


class ContextSectionType(str, Enum):
    PROJECT_BRIEF = "project_brief"
    WORLD = "world"
    LORE = "lore"
    CHARACTERS = "characters"
    PLOT_HOOKS = "plot_hooks"
    CHAPTER_OUTLINES = "chapter_outlines"
    MAP_REGIONS = "map_regions"
    WORKFLOW_STATE = "workflow_state"
    WRITING_RULES = "writing_rules"
    AGENT_TEMPLATES = "agent_templates"


class AssistantContextBudget(BaseModel):
    max_context_tokens: int = Field(default=16000, ge=1000)
    reserved_response_tokens: int = Field(default=4000, ge=0)
    history_tokens: int = Field(default=2500, ge=0)
    snapshot_tokens: int = Field(default=9000, ge=0)
    delta_tokens: int = Field(default=1500, ge=0)
    retrieval_tokens: int = Field(default=3000, ge=0)


class AssistantContextSessionCreateRequest(BaseModel):
    assistant_surface: AssistantSurface
    mode: str = "default"
    session_id: Optional[str] = None
    scope: Dict[str, Any] = Field(default_factory=dict)
    request_id: Optional[str] = None


class AssistantHistoryResetRequest(BaseModel):
    assistant_surface: AssistantSurface
    mode: str = "default"
    clear_pending_items: bool = False
    delete_message_records: bool = True
    reason: str = "user_requested"
    request_id: Optional[str] = None


class AssistantForceRereadRequest(BaseModel):
    assistant_surface: AssistantSurface
    session_id: Optional[str] = None
    mode: str = "default"
    scope: Dict[str, Any] = Field(default_factory=dict)
    clear_history: bool = False
    clear_pending_items: bool = False
    reason: str = "user_requested_after_project_update"
    request_id: Optional[str] = None


class AssistantPacketBuildRequest(BaseModel):
    assistant_surface: AssistantSurface
    task_type: str = "chat"
    session_id: Optional[str] = None
    mode: str = "default"
    request_id: Optional[str] = None
    scope: Dict[str, Any] = Field(default_factory=dict)
    budget: AssistantContextBudget = Field(default_factory=AssistantContextBudget)
    user_message: Optional[str] = None
    force_reread: bool = False
    reset_history: bool = False


class AssistantContextSummary(BaseModel):
    packet_id: Optional[str] = None
    snapshot_id: Optional[str] = None
    snapshot_version: Optional[int] = None
    assistant_surface: Optional[str] = None
    invalidation_state: str = PacketInvalidationState.FRESH.value
    force_reread: bool = False
    history_reset_applied: bool = False
    token_estimate: int = 0
    selected_sections: List[Dict[str, Any]] = Field(default_factory=list)
    omitted_sections: List[Dict[str, Any]] = Field(default_factory=list)
    delta_count: int = 0
    rebuilt_at: Optional[datetime] = None


class AssistantSessionResponse(BaseModel):
    session_id: str
    project_id: str
    assistant_surface: str
    mode: str = "default"
    status: str = AssistantSessionStatus.ACTIVE.value
    snapshot_id: Optional[str] = None
    snapshot_version: Optional[int] = None
    history_summary: str = ""
    history_window: List[Dict[str, Any]] = Field(default_factory=list)
    pending_items: List[Dict[str, Any]] = Field(default_factory=list)
    context_cursor: Dict[str, Any] = Field(default_factory=dict)
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class AssistantMessageResponse(BaseModel):
    id: str
    session_id: str
    role: str
    content: str
    request_id: Optional[str] = None
    metadata: Dict[str, Any] = Field(default_factory=dict)
    packet_id: Optional[str] = None
    snapshot_id: Optional[str] = None
    created_at: Optional[datetime] = None


class AssistantHistoryResponse(BaseModel):
    session: AssistantSessionResponse
    messages: List[AssistantMessageResponse] = Field(default_factory=list)
    context_packet: Optional[AssistantContextSummary] = None


class AssistantHistoryResetResponse(BaseModel):
    success: bool = True
    old_session_id: str
    new_session_id: str
    history_reset_at: datetime = Field(default_factory=datetime.utcnow)
    messages_deleted: int = 0
    messages_archived: int = 0
    pending_items_cleared: bool = False
    snapshot_id: Optional[str] = None
    snapshot_version: Optional[int] = None


class AssistantForceRereadResponse(BaseModel):
    success: bool = True
    project_id: str
    session_id: Optional[str] = None
    snapshot_id: str
    snapshot_version: int
    previous_snapshot_id: Optional[str] = None
    previous_snapshot_version: Optional[int] = None
    history_cleared: bool = False
    force_reread: bool = True
    packet_metadata: AssistantContextSummary


class AssistantSnapshotStatusResponse(BaseModel):
    project_id: str
    snapshot_id: Optional[str] = None
    snapshot_version: Optional[int] = None
    status: str = SnapshotStatus.STALE.value
    source_revision_hash: Optional[str] = None
    content_hash: Optional[str] = None
    token_estimate: int = 0
    stale_delta_count: int = 0
    build_reason: Optional[str] = None
    built_at: Optional[datetime] = None
    error_message: Optional[str] = None


class AssistantPacketResponse(BaseModel):
    packet_id: str
    project_id: str
    session_id: Optional[str] = None
    assistant_surface: str
    request_id: Optional[str] = None
    snapshot_id: Optional[str] = None
    snapshot_version: Optional[int] = None
    packet_scope: Dict[str, Any] = Field(default_factory=dict)
    budget: Dict[str, Any] = Field(default_factory=dict)
    selected_sections: List[Dict[str, Any]] = Field(default_factory=list)
    delta_ids: List[str] = Field(default_factory=list)
    retrieval_manifest: Dict[str, Any] = Field(default_factory=dict)
    token_estimate: int = 0
    truncated: bool = False
    invalidation_state: str = PacketInvalidationState.FRESH.value
    force_reread: bool = False
    history_reset_applied: bool = False
    metadata: Dict[str, Any] = Field(default_factory=dict)
    content_hash: Optional[str] = None
    created_at: Optional[datetime] = None


def new_assistant_session_id(surface: str) -> str:
    prefix = "as"
    safe_surface = "".join(ch for ch in surface if ch.isalnum())[:8] or "session"
    return f"{prefix}_{safe_surface}_{uuid4().hex[:12]}"
